File: players.py
from copy import deepcopy

from board import Board
from minimax import minimax
from heuristics import Heuristic

PLAYER_MAPPINGS = {
    0: "Black",
    1: "White",
    2: "Red",
}


class Player:
    def __init__(self, turn):
        self.turn = turn
        self.color = PLAYER_MAPPINGS[turn]

    def print_board(self, board: Board) -> None:
        raise NotImplementedError

    def print_no_moves(self) -> None:
        raise NotImplementedError

    def get_move(self, game) -> (int, int):
        raise NotImplementedError

    def has_valid_move(self, board: Board) -> bool:
        return board.has_valid_move(self.turn)


class HumanPlayer(Player):
    def print_board(self, board: Board) -> None:
        print(board)
        print(f"{self.color}'s turn.")

    def print_no_moves(self) -> None:
        print(f"{self.color} has no moves, skipping.")

    def get_move(self, game) -> (int, int):
        move = input("Provide coords as '<int>, <int>': ")
        # loop till valid move
        while True:
            try:
                row, col = map(int, move.split(",", maxsplit=1))
            except ValueError:
                move = input("Invalid input string. Try again: ")
                continue
            valid, reason = game.board.validate_placing(row, col, self.turn)
            if not valid:
                move = input(f"{reason} Try again: ")
            else:
                return row, col


class AIPlayer(Player):
    def __init__(self, turn, heuristic: Heuristic) -> None:
        super().__init__(turn)
        self.heuristic = heuristic

    def print_board(self, board: Board) -> None:
        pass  # Robots don't need visuals

    def print_no_moves(self) -> None:
        pass  # Robots don't need visuals

    def get_move(self, game) -> (int, int):
        best_eval = None
        best_move = None
        valid_moves = game.board.valid_moves(self.turn)
        next_player = (self.turn + 1) % len(game.players)

        for move_row, move_col in valid_moves:
            # TODO(tkarwowski): making this backtrace instead of coping breaks everything
            new_board = deepcopy(game.board)
            new_board.place(move_row, move_col, self.turn)
            evaluation = minimax(
                new_board,
                game.minimax_depth - 1,
                self.turn,
                next_player,
                len(game.players),
                0,
                self.heuristic.evaluate,
                None,
                None,
            )
            best_eval = evaluation if best_eval is None else max(best_eval, evaluation)
            if best_eval == evaluation:
                best_move = (move_row, move_col)

        # TODO(tkarwowski): temporary debug statement (but looks cool)
        # print(game.board)
        return best_move

File: src/players.rs
use std::io;
use std::io::Write;

use crate::algorithms::minimax;
use crate::board::{BoardState, Move};

const NUM_PLAYERS: usize = 3;

pub(crate) trait Player {
    fn get_move(&self, board: &BoardState, turn: usize) -> (usize, usize);
    fn print_board(&self, board: &BoardState);
    fn print_no_moves(&self);
}

pub(crate) struct Human {}

impl Human {
    pub(crate) fn new() -> Self {
        Human {}
    }
}

impl Player for Human {
    fn get_move(&self, board: &BoardState, player: usize) -> (usize, usize) {
        let mut input = String::with_capacity(4);
        print!("Provide coords as '<int>, <int>': ");
        // loop till valid move is provided
        loop {
            io::stdout().flush().unwrap();
            input.clear();
            io::stdin().read_line(&mut input).unwrap();
            let (row, col) = if let Some((row, col)) = input.split_once(",") {
                match (
                    row.trim().parse(),
                    col.trim_end_matches('\n').trim().parse(),
                ) {
                    // we're only interested if everything is parsed correctly
                    (Ok(row), Ok(col)) => (row, col),
                    (_, _) => {
                        print!("Can't parse input. Try again: ");
                        continue;
                    }
                }
            } else {
                print!("Expected two arguments separated by comma. Try again: ");
                continue;
            };
            match board.validate_placing(row, col, player) {
                Move::Valid { .. } => return (row, col),
                Move::Invalid(reason) => {
                    print!("{} Try again: ", reason);
                }
            }
        }
    }

    fn print_board(&self, board: &BoardState) {
        print!("{}", board);
        // println!("{}'s turn.", "nobody");
    }

    fn print_no_moves(&self) {
        println!("No available moves. Skipping turn.");
    }
}

pub(crate) struct AiPlayer<F>
where
    F: Fn(&BoardState, usize) -> i32,
{
    minimax_depth: usize,
    heuristic: F,
}

impl<F> AiPlayer<F>
where
    F: Fn(&BoardState, usize) -> i32,
{
    pub(crate) fn new(minimax_depth: usize, heuristic: F) -> Self {
        AiPlayer {
            minimax_depth,
            heuristic,
        }
    }
}

impl<F> Player for AiPlayer<F>
where
    F: Fn(&BoardState, usize) -> i32,
{
    fn get_move(&self, board: &BoardState, player: usize) -> (usize, usize) {
        minimax(
            board,
            &self.heuristic,
            player,
            NUM_PLAYERS,
            self.minimax_depth,
        )
    }

    fn print_board(&self, board: &BoardState) {
        print!("{}", board);
    }

    fn print_no_moves(&self) {
        println!("No available moves. Skipping turn.");
    }
}
